core/templates/components/button-directives/create-activity-button.component.spec.ts

```diff
@@ -0,0 +1,291 @@
+    isLoggedIn: () => true
+  };
+
+  let userInfoForNonCollectionCreator = {
+    _isModerator: true,
+    _isAdmin: false,
+    _isTopicManager: false,
+    _isSuperAdmin: false,
+    _canCreateCollections: true,
+    _preferredSiteLanguageCode: 'en',
+    _username: 'username1',
+    _email: 'tester@example.org',
+    _isLoggedIn: true,
+    isModerator: () => true,
+    isAdmin: () => false,
+    isSuperAdmin: () => false,
+    isTopicManager: () => false,
+    canCreateCollections: () => false,
+    getPreferredSiteLanguageCode: () =>'en',
+    getUsername: () => 'username1',
+    getEmail: () => 'tester@example.org',
+    isLoggedIn: () => true
+  };
+
+  beforeEach(waitForAsync(() => {
+    windowRef = new MockWindowRef();
+    TestBed.configureTestingModule({
+      imports: [HttpClientTestingModule],
+      declarations: [
+        CreateActivityButtonComponent,
+        TranslatePipe
+      ],
+      providers: [
+        UserService,
+        {
+          provide: UrlService,
+          useClass: MockUrlService
+        },
+        {
+          provide: WindowRef,
+          useValue: windowRef
+        },
+        ExplorationCreationService
+      ]
+    }).compileComponents();
+  }));
+
+  beforeEach(() => {
+    fixture = TestBed.createComponent(CreateActivityButtonComponent);
+    component = fixture.componentInstance;
+    userService = TestBed.inject(UserService);
+    urlService = TestBed.inject(UrlService);
+    explorationCreationService = TestBed.inject(ExplorationCreationService);
+    ngbModal = TestBed.inject(NgbModal);
+    fixture.detectChanges();
+  });
+
+  it('should successfully instantiate the component from beforeEach block',
+    () => {
+      expect(component).toBeDefined();
+    });
+
+  it('should begin activity creation process if user can' +
+    ' create collections', fakeAsync(() => {
+    spyOn(userService, 'getUserInfoAsync')
+      .and.returnValue(Promise.resolve(userInfoForCollectionCreator));
+
+    spyOn(component, 'initCreationProcess');
+
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(false);
+
+    component.ngOnInit();
+    tick();
+    fixture.detectChanges();
+
+    expect(component.initCreationProcess).toHaveBeenCalled();
+    expect(component.canCreateCollections).toBe(true);
+    expect(component.userIsLoggedIn).toBe(true);
+  }));
+
+  it('should create a new exploration automatically if the user' +
+    ' cannot create collections', fakeAsync(() => {
+    spyOn(userService, 'getUserInfoAsync')
+      .and.returnValue(Promise.resolve(userInfoForNonCollectionCreator));
+    const explorationCreationServiceSpy = spyOn(
+      explorationCreationService, 'createNewExploration');
+
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(false);
+
+    component.ngOnInit();
+    tick();
+    fixture.detectChanges();
+
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
```

Comment:
Nit: Conceptually, it may make more sense for these three expectations to be in reverse order: First, the user is logged in. Then, we find that they can't create collections. Finally, we confirm that the exploration creation service has been called.

I.e. try to have  each test tell a coherent story. That makes the test easier to follow.

Reply:
Yep, that sounds better. Changed